File: Ex02/zurg.pl
#!/usr/bin/perl
# -*-Perl-*-
# Last changed Time-stamp: <2008-05-09 14:37:30 xtof>
# $Id$

use Data::Dumper;
use Getopt::Long;
use Pod::Usage;
use strict;
use vars qw/$DEPTH $EXHAUSTIVE $MAXCOST %STATE %TOYS $VERBOSE/;
use warnings;

# defaults for global(s)
%TOYS = ( Buzz  =>  5,
	  Woody => 10,
	  Rex   => 20,
	  Hamm  => 25,
	  xtof  => 30,
	  ivo   =>  7 );

# initial state
%STATE = ( history  => [],
	   left     => [keys %TOYS],
	   right    => [],
	   cost     => 0,
	   lightpos => 'L' );

# initial maximal cost
$MAXCOST = 9999999;

# toggles
$EXHAUSTIVE = 0;
$VERBOSE = 0;

# process command-line
pod2usage(-verbose => 0)
    unless GetOptions(
      "toys=s"     => sub { %TOYS = split(':', $_[1]);
			    $STATE{left} = [keys %TOYS] },
      "exhaustive" => \$EXHAUSTIVE,
      "help"       => sub{pod2usage(-verbose => 1)},
      "man"        => sub{pod2usage(-verbose => 2)},      
      "maxcost=i"  => \$MAXCOST,
      "verbose"    => sub{ $VERBOSE++ }
      );

# main
print "Optimal cost is ", solve(\%STATE), "\n";
print Dumper(\%STATE);

#---
sub solve {
  my $s = shift;
  $DEPTH++;
  die if $DEPTH>20;
  print "$DEPTH: @{$s->{left}} | @{$s->{right}} $s->{cost}\n" if $VERBOSE >=2;

  $DEPTH--, return 0 unless @{$s->{left}};
  my $mvs = moves($s);
  my $best = 999999999;
  my $curr = $s->{cost};
  my %bstate;
  
  foreach my $m (@$mvs) {
    unless ($EXHAUSTIVE) {
      next if $m->[0] + $curr > $MAXCOST;
    }
    my %state = apply_move($s, $m);
    my $cost = $m->[0] + solve(\%state);
    print "$DEPTH: @$m + @{[$cost-$m->[0]]} = $cost\n" if $VERBOSE;
    $best = $cost, %bstate = %state if $cost<$best;
    $MAXCOST = $best + $curr if $MAXCOST > $best + $curr;
  }
  
  %$s = %bstate;
  $DEPTH--;
  return $best;
}

#---
# state, move -> updated_state
#---
sub apply_move {
  my %state =  %{shift()};
  my @m = @{shift()};
  my ($from, $to) = ($state{lightpos} eq 'L')
                  ? ('left','right')
		  : ('right','left');

  # update cost
  $state{cost} += shift @m;

  # update toys left/right of bridge according to move
  foreach my $t (@m) {
    $state{$from} = [ grep {! m/$t/} @{$state{$from}} ];
    $state{$to}   = [ @{$state{$to}}, $t ];
  }

  # update flashlight position
  $state{lightpos} = ($state{lightpos} eq 'L') ? 'R' : 'L';
  # update move history
  $state{history}  = [ @{$state{history}}, [ @m ] ];

  return %state;
}

#---
# state -> [move1, move2, ..., moveN]
#---
sub moves {
  my $state = shift();
  my @moves = ();

  if ($state->{'lightpos'} eq 'L') {
    # flashlight is on the left, therefore all subsets of cardinality
    # two of the set of toys left of the bridge form the legal moves.
    push @moves, all_pairs($state->{'left'});
  }
  else {
    # flashlight is on the right, therefore all subsets of cardinality
    # one of the set of toys right of the bridge form the legal moves.
    push @moves, map {[$TOYS{$_},$_]} @{$state->{right}};
  }

  return \@moves;
}

#---
# generates all combinations of 2 out of N array elements.
# ---
sub all_pairs {
  my $array = shift;
  my @combis = ();
  foreach my $idx1 (0 .. $#$array) {
    my $t1 = $array->[$idx1];
    foreach my $t2 (@{$array}[$idx1+1..$#$array]) {
      push @combis, [max($TOYS{$t1}, $TOYS{$t2}), $t1, $t2];
    }
  }

  return @combis;
}

#---
sub max {
  return ($_[0]>$_[1]) ? $_[0] : $_[1];
}

=pod

=head1 NAME

zurg.pl - solves "Escape from Zurg" problems

=head1 SYNOPSIS

zurg.pl [[-exhaustive] [-maxcost I<INT>] [-man] [-help] [-toys I<STRING>]]

=head1 OPTIONS

=over 4

=item B<-exhaustive>

Make search exhaustive.

=item B<-help>

Show synopsis.

=item B<-maxcost> I<INT>

Set maximal cost of solution to I<INT>.

=item B<-man>

Show man page.

=item B<-toys> I<STRING>

Set toys list. e.g.  -toys Buzz:5:Woody:10:Rex:20:Hamm:25

=item B<-verbose>

Toggle on verbose mode.
    
=back
    
=head1 DESCRIPTION

The "Escape from Zurg" problem reads as follows:

Buzz, Woody, Rex, and Hamm have to escape from Zurg.a They merely have
to cross one last bridge before they are free. However, the bridge is
fragile and can hold at most two of them at the same time. Moreover,
to cross the bridge a flashlight is needed to avoid traps and broken
parts. The problem is that our friends have only one flashlight with
one battery that lasts for only 60 minutes (this is not a typo:
sixty). The toys need different times to cross the bridge (in either
direction):

 +-------+------------+
 |  TOY  |    TIME    | 
 +=======+============+
 | Buzz  |  5 minutes |
 +-------+------------+
 | Woody | 10 minutes |
 +-------+------------+
 | Rex   | 20 minutes |
 +-------+------------+
 | Hamm  | 25 minutes |
 +-------+------------+
    
Since there can be only two toys on the bridge at the same time, they
cannot cross the bridge all at once. Since they need the flashlight to
cross the bridge, whenever two have crossed the bridge, somebody has
to go back and bring the flashlight to those toys on the other side
that still have to cross the bridge.  The problem now is: In which
order can the four toys cross the bridge in time (that is, in 60
minutes) to be saved from Zurg?

=head1 AUTHORS

Ivo Hofacker, Christoph Flamm
    
=head1 LITERATURE

 Martin Erwig "Escape from Zurg: an exercise in logic programming",
 J Func Prog, 14:(3) 253-261 (2004); doi: 10.1017/S0956796804005040

=cut

__END__

File: Ex02/zurg.c
/* Last changed Time-stamp: <2009-03-31 14:40:36 ivo> */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

#define MAX2(A,B)  ((A)>(B)?(A):(B))

static char rcsid[] = "$Id:  $";

enum toys {
  Buzz,
  Woody,
  Rex,
  Hamm,
  xtof,
  ivo,
  Speedy,
  toynum
};


enum light {
  left,
  right
};

char *names[] = { "Buzz", "Woody", "Rex", "Hamm", "xtof", "ivo", "Speedy" };
int costs[] = {5, 10, 20, 25, 30, 7, 2};


struct state {
  int cost;
  enum light lightpos;
  unsigned int where;  /* bitvector encoding where each toy is */
  enum toys history[3*toynum];
  int histlen;
};

struct move {
  enum toys first, second;
  int cost;
};

struct move *moves(struct state *s);
struct state move_it(struct state *s, struct move *m);

static int maxcost = INT_MAX;

int solve(struct state *s) {
  struct move *mvs, *m;
  int best, curr;
  struct state bstate;

  if (s->where +1 == 1 << toynum) {
    return 0;
  }
  mvs = moves(s);
  best = INT_MAX;
  curr = s->cost;
  for (m=mvs; m->cost>0; m++) {
    int cost;
    struct state test, gstate;
    if (curr + m->cost > maxcost) continue;
    gstate = test = move_it(s, m);
    if (greedy(&gstate) > maxcost) continue;
    cost = m->cost + solve(&test);
    if (best>cost) {
      best = cost;
      bstate = test;
    }
    if (maxcost < best + curr) maxcost = best + curr;
  }
  free(mvs);
  *s = bstate;
  return best;
}

struct move *moves(struct state *s) {
  struct move *m;
  int num=0;
  enum toys i,j;

  m = calloc(sizeof(struct move),(toynum*(toynum+1))/2);

  if (s->lightpos==left) {
    for (i=0; i<toynum; i++) {
      if (s->where & 1 << i) continue;
      for (j=i+1; j<toynum; j++) {
	if (s->where & 1 << j) continue;
	m[num].first = i;
	m[num].second = j;
	m[num++].cost = MAX2(costs[i],costs[j]);
      }
    }
  } else {
    for (i=0; i<toynum; i++) {
      if (s->where & 1 << i) {
	m[num].first = m[num].second = i;
	m[num++].cost = costs[i];
      }
    }
  }
  return m;
}

struct state move_it(struct state *s, struct move *m) {
  struct state new;

  new = *s;
  new.where ^= 1 << m->first;
  new.history[new.histlen++] = m->first;
  new.cost += m->cost;
  if (s->lightpos == left) {
    new.where ^= 1 << m->second;
    new.history[new.histlen++] = m->second;
    new.lightpos = right;
  } else
    new.lightpos = left;
  return new;
}

int greedy(struct state *s) {
  struct move *mvs, *m, *best;

  while (s->where +1 != 1 << toynum) {

    mvs = moves(s);
    best = mvs;
    for (m=mvs+1; m->cost>0; m++) {
      if (m->cost < best->cost)
	best = m;
    }
    
    *s = move_it(s, best);
    free(mvs);
  }
  return s->cost;
}

void printSolution(struct state *s) {
  int i;
  
  printf("Solution with cost\t %2d:\n", s->cost);
  printf("%-5s, %-5s cross the bridge\t(cost %2d)\n",
	 names[s->history[0]], names[s->history[1]],
	 MAX2(costs[s->history[0]],costs[s->history[1]]));
  for (i=2; i<s->histlen; i+=3) {
    printf("%-5s\t     goes back\t\t(cost %2d)\n",
	   names[s->history[i]], costs[s->history[i]]);
  printf("%-5s, %-5s cross the bridge\t(cost %2d)\n",
	 names[s->history[i+1]], names[s->history[i+2]],
	 MAX2(costs[s->history[i+1]],costs[s->history[i+2]]));
  }
  printf("\n");
}

int DP(void) {
  



}

int main() {
  struct state state, gstate;

  state.lightpos=left;
  state.where=0;
  state.histlen=0;
  state.cost=0;

  gstate = state;

  printf("Greedy ");
  maxcost = greedy(&gstate);
  printSolution(&gstate);

  printf("Optimal ");
  solve(&state);
  printSolution(&state);

  exit(EXIT_SUCCESS);
}
